Added docsify docs (#411)
airbnb/MvRx: docs/counter.md

# Mavericks: Counter App

Getting started with Mavericks is simple. Once it is [integrated into your project](setup.md), let's walk through a simple counter example. We're going to make this:

![Counter](/images/counter.png)

When you click the counter text, the count goes up by one.
Code for this sample can be found [here](https://github.com/airbnb/MvRx/tree/master/counter).

## Create CounterState

The first step in creating a Mavericks screen is to model it as a function of state. Modeling a screen as a function of state is a useful concept because it is:
1. Easy to test

> 1. Easy for you and other engineer to reason through

Reviewer: ```suggestion
1. Easy for you and other engineers to reason through
```

1. Renders the same independently of the order of events leading up to it
1. Powerful enough to render any type of screen

A counter screen is simple to model.
```kotlin
data class CounterState(val count: Int = 0) : MavericksState
```
* CounterState contains a single immutable (`val`) property, count, it is an integer, and it defaults to 0.
* CounterState is a [data class](https://kotlinlang.org/docs/reference/data-classes.html) so that Kotlin creates equals and hashCode automatically.

> * CounterState implements `MavericksState` signals the intention that this class will be used as Mavericks state.

Reviewer: ```suggestion
* CounterState implementation of `MavericksState` signals the intention that this class will be used as Mavericks state.
```

## Create CounterViewModel

Next, we need to create a ViewModel which will own CounterState and be responsible for updating it.
```kotlin
class CounterViewModel(initialState: CounterState) : MavericksViewModel<CounterState>(state) {
    fun incrementCount() {
      setState { copy(count = count + 1) }
    }
}
```
* CounterViewModel takes CounterState as a constructor parameter because Mavericks creates your initial state for you. It does this to handle [saved state](/saved-state.md) across process restoration.

> * incrementCount() calls setState which takes a reducer with the type `S.() -> S`. The receiver of the reducer is the current state when the reducer is run and it returns the updated state. `copy` is the kotlin data class copy function.

Reviewer: ```suggestion
* incrementCount() calls setState which takes a reducer with the type `CounterState.() -> CounterState `. The receiver of the reducer is the current state when the reducer is run and it returns the updated state. `copy` is the kotlin data class copy function.
```
